libsignal-protocol/src/crypto.rs

```diff
@@ -56,49 +299,185 @@ impl CryptoProvider {
+    user_data.0.hmac_sha256_cleanup();
 }
 
 unsafe extern "C" fn hmac_sha256_final_func(
     _hmac_context: *mut c_void,
-    _output: *mut *mut signal_buffer,
-    _user_data: *mut c_void,
+    output: *mut *mut signal_buffer,
+    user_data: *mut c_void,
 ) -> i32 {
-    unimplemented!()
+    assert!(!user_data.is_null());
+
+    let user_data = &mut *(user_data as *mut State);
+    match user_data.0.hmac_sha256_final() {
+        Ok(buf) => {
+            let buffer = Buffer::from(buf);
+            output.write(buffer.into_raw());
+            sys::SG_SUCCESS as c_int
+        },
+        Err(e) => e.code(),
+    }
 }
 
 unsafe extern "C" fn hmac_sha256_init_func(
     _hmac_context: *mut *mut c_void,
-    _key: *const u8,
-    _key_len: usize,
-    _user_data: *mut c_void,
+    key: *const u8,
+    key_len: usize,
+    user_data: *mut c_void,
 ) -> i32 {
-    unimplemented!()
+    assert!(!key.is_null());
+    assert!(!user_data.is_null());
+
+    let user_data = &mut *(user_data as *mut State);
+    let buffer = slice::from_raw_parts(key, key_len);
+    user_data.0.hmac_sha256_init(buffer).into_code()
 }
 
 unsafe extern "C" fn hmac_sha256_update_func(
     _hmac_context: *mut c_void,
-    _data: *const u8,
-    _data_len: usize,
-    _user_data: *mut c_void,
+    data: *const u8,
+    data_len: usize,
+    user_data: *mut c_void,
 ) -> i32 {
-    unimplemented!()
+    assert!(!data.is_null());
+    assert!(!user_data.is_null());
+
+    let user_data = &mut *(user_data as *mut State);
+    let buffer = slice::from_raw_parts(data, data_len);
+    user_data.0.hmac_sha256_update(buffer).into_code()
 }
 
-unsafe extern "C" fn random_func(
-    data: *mut u8,
-    len: usize,
+unsafe extern "C" fn sha512_digest_init_func(
+    _digest_context: *mut *mut c_void,
+    user_data: *mut c_void,
+) -> c_int {
+    assert!(!user_data.is_null());
+
+    let user_data = &mut *(user_data as *mut State);
+    user_data.0.sha512_digest_init().into_code()
+}
+
+unsafe extern "C" fn sha512_digest_update_func(
+    _digest_context: *mut c_void,
+    data: *const u8,
+    data_len: usize,
     user_data: *mut c_void,
 ) -> c_int {
     assert!(!data.is_null());
     assert!(!user_data.is_null());
 
-    let user_data = &*(user_data as *const State);
-    let buffer = slice::from_raw_parts_mut(data, len);
-    user_data.0.fill_random(buffer);
-    0
+    let user_data = &mut *(user_data as *mut State);
+    let buffer = slice::from_raw_parts(data, data_len);
+    user_data.0.sha512_digest_update(buffer).into_code()
+}
+
+unsafe extern "C" fn sha512_digest_final_func(
+    _digest_context: *mut c_void,
+    output: *mut *mut signal_buffer,
+    user_data: *mut c_void,
+) -> c_int {
+    assert!(!user_data.is_null());
+
+    let user_data = &mut *(user_data as *mut State);
+    match user_data.0.sha512_digest_final() {
+        Ok(buf) => {
+            let buffer = Buffer::from(buf);
+            output.write(buffer.into_raw());
```

Comment:
Like the previous `ptr.write()` comment, would it be more readable if we write this as `*output = buffer.into_raw()`?

Reply:
Yeah, with check of course for `assert!(! output.is_null())`.